File: lib/dbg.rb

require "callerid"
require "fileutils"
require "htmlentities"
require 'sysutils'
require "wildcardlookup"

class DbgMgr
   prohibit_subclassing!

private
   def self.raise_NoCopyError(method_name)
      msg = "DbgMgr is intended for use as a singleton: '#{method_name}' is not allowed!"
      raise NotImplementedError, msg
   end
public


   def self.new(*args) 
      raise NotImplementedError, 'DbgMgr has no instances!'
   end

   def self.clone(freeze: true)
      raise_NoCopyError("clone")
   end

   def self.dup()
      raise_NoCopyError("dup")
   end

   class << self
      attr_reader :esc, # if true, escape HTML entities in the output (esc is false by default)
         :path,      # if non-nil, the file to which to write the buffer
         :styles,    # user-added CSS definitions
         :count,     # a cumulative count of messages written--not the buffer.length
         :group_by   # how to group the messages before sorting by time of arrival.
      private
      attr_accessor :active,  # the "set" of live keys
         :important, # a subset of the active flags that deserved to be seen first
         :buffer     # when buffered debugging is done, the list of messages
      attr_writer :path, # if non-nil, the file to which to write the buffer
         :styles,    
         :count      
   end
   @active = WildCardLookup.new
   @important = WildCardLookup.new

   @buffer = []
   @count = 0
   @group_by = :time

   @path = nil
   @styles = ''
   @esc = false
   MESSAGE_GROUPS = Set.new %i[caller key match priority time] # for grouping buffered message sets

   def self.add_patterns(*args)
      args.each do |arg|
         if arg.kind_of? String
            raw_patterns = arg.split ','
            raw_patterns.each do |pattern|
               if pattern[0] == '+'
                  real_pattern = pattern[1 .. -1]
                  @active.add_pattern real_pattern
                  @important.add_pattern real_pattern
               elsif pattern[0] == '@'
                  group = pattern[1 .. -1]
                  found = MESSAGE_GROUPS.find { |name| name.to_s.begin_with? group}
                  if found.nil?
                     self.warn("DbgMgr.add_patterns", "unknown message group: '#{group}'")
                     @active.add_pattern pattern
                  else
                     @group_by = found
                  end
               else
                  @active.add_pattern pattern
               end
            end # do |pattern|
         elsif arg.kind_of? Array
            add_patterns *arg
         else 
            raise ArgumentError.new "Unexpected argument type, #{arg}"
         end
      end
      nil 
   end

   def self.remove_patterns(*args)
      args.each do |arg|
         if arg.is_a? String
            patterns = arg.split(',')
            @active.remove_patterns patterns
            @important.remove_patterns patterns
         elsif arg.is_a? Array 
            remove_patterns *arg
         else
            raise ArgumentError.new "Unexpected argument type, #{arg}"
         end
      end
   end


   def self.active?(*args)
      args.each do |arg|
         if arg.is_a? String 
            all = arg.split(',')
            return @active.any_matched? *all
         elsif arg.is_a? Array and active? *arg then return true
         else raise ArgumentError.new "Unexpected argument type, #{arg}"
         end
      end
      false
   end

   def self.all_active?(*args)
      args.each do |arg|
         if arg.is_a? String and not @active.all_matched? arg.split(',') then return false
         elsif arg.is_a? Array and not all_active? *arg then return false
         else raise ArgumentError.new "Unexpected argument type, #{arg}"
         end
      end
      true
   end

   def self.open(path='.', asis: false, esc: false)
      abspath = File.join File.expand_path path
      if not asis and not abspath.end_with? "#{File::SEPARATOR}log"
         abspath = File.join abspath, "#{File::SEPARATOR}log"
      end
      if @path == abspath then return self # caller knows last file is the only one! 
      elsif not @path.nil?
         msg = "Already logging to #{@path}, cannot use #{abspath}"
         STDERR.puts msg
         warn msg
         self
      end
      File.assure_directory(abspath)
      @path = File.join abspath, "dbg#{Time.now.strftime '%y%m%d_%H%M%S'}.html"
      File.open(@path, mode: 'w') do |stream|
         stream.puts "<!DOCTYPE html>\n<html>\n<head>"
         stream.puts "   <meta charset=\"utf-8\">"
         stream.puts "   <meta name=\"generator\" content=\"rubymd\">"
         stream.puts "   <style type=\"text/css\">"
         stream.puts DbgMessage::STYLES
         stream.puts ADMIN_STYLES
         stream.puts " #{@styles}\n    </style>\n</head>\n<body>\n<code>" 
      end
      @esc = esc
      @styles = ''
      self
   end
   def self.add_important_patterns(*args)
      args.each do |arg|
         if arg.is_a? String
            add_patterns arg
            @important.add_patterns arg.split(',')
         elsif arg.is_a? Array 
            add_important_patterns *arg
         else
            raise ArgumentError.new "Unexpected argument type: #{arg.class}"
         end
      end
   end

   def self.not_important(*args)
      args.each do |arg|
         if arg.is_a? String
            @important.remove_patterns arg.split(',')
         elsif arg.is_a? Array 
            not_important *arg
         else
            raise ArgumentError.new "Unexpected argument type: #{arg.class}"
         end
      end
   end
   def self.esc= yes_or_no
      @esc = yes_or_no.to_b
   end
 
   @encoder = HTMLEntities.new 
   def self.html_encode text
      @encoder.encode text, :named
   end

   def self.put(key, text, 
      esc: @esc, my_caller: CallerId.my_caller, style: nil
   )
      if key != "dbgmgr" and @active.matched?("dbgmgr")
         tgt = if @path.nil? then "STDOUT" else @path end
         self.put "dbgmgr", "put #{key}? #{@active.matched? key}, target #{tgt}"
      end
      if not @active.matched? key
         nil
      elsif @path.nil?
         puts text
         text
      else
         matcher = @active.weakest_pattern key
         if esc then text = html_encode text end
         @count += 1
         msg = if @important.matched? key
                  DbgMessage.priority(
                     key: key, match: matcher, index: @count, 
                     caller: my_caller, body: text, style: style
                  )
               else
                  DbgMessage.ordinary(
                     key: key, match: matcher, index: @count,
                     caller: my_caller, body: text, style: style
                  )
               end
         @buffer.push msg
         msg
      end
   end
   def self.err(error, esc: @esc, my_caller: CallerId.my_caller, now: false, bail_out: true)
      if error.respond_to? :backtrace
         backtrace  = (error.backtrace or []).join('\n')
      else
         backtrace = ''
      end
      class_name = if error.kind_of? Exception then "#{error.class}: " else "" end
      if @path.nil?
         STDERR.puts "#{class_name}#{error}"
         if backtrace.length > 0
            STDERR.puts backtrace
         end
      else
         @count += 1
         text = class_name + error.to_s
         if esc then text = html_encode text end
         if backtrace.length > 0
            text += "\nTraceback:\n#{html_encode backtrace}"
         end
         final_text = "<pre>#{text}</pre>"
         msg = DbgMessage.err index: @count, caller: my_caller, body: final_text
         @buffer.push msg
         if now 
            if now.is_a? String then flush after: now
            else DbgMgr.flush
            end
         end
      end
      if bail_out
         if class_name.length > 0
            raise error
         else
            raise RuntimeError(error.to_s)
         end
      else
         msg
      end
   end
   def self.pre(key, text, esc: @esc, my_caller: CallerId.my_caller, style: nil)
      # If the caller wants us to escape, we have to do it here so that our "<pre>"  and
      # "</pre>" do not get escaped later.
      if active? key
         if (@path != nil) and esc then text = html_encode text end
         put key, "<pre>#{text}</pre>", esc: false, my_caller: my_caller
      end
   end  
   def self.warn(warning, esc: nil, my_caller: CallerId.my_caller, now: false, bail_out: false)
      if warning.respond_to? :backtrace
         backtrace  = (warning.backtrace or []).join('\n')
      else
         backtrace = ''
      end
      if @path.nil?
         STDERR.puts warning
         if backtrace.length > 0
            STDERR.puts backtrace
         end
         warning
      else
         @count += 1
         warning = if esc then html_encode warning.to_s else warning.to_s end
         if backtrace.length > 0
            warning += "\nTraceback:\n#{html_encode backtrace}"
         end
         msg = DbgMessage.warn index: @count, caller: my_caller, body: warning
         @buffer.push msg
         if now
            if now.is_a? String then flush after: now
            else now DbgMgr.flush
            end
         end
      end
      if bail_out
         if class_name.length > 0
            raise error
         else
            raise RuntimeError(error.to_s)
         end
      else
         msg
      end
   end
   def self.group_by=(field)
      as_symbol = field.to_sym
      if MESSAGE_GROUPS.member? as_symbol
         @group_by = as_symbol
      else
         raise ArgumentError.new "Unrecognized message group, '#{field}'" 
      end
   end


   ADMIN_STYLES = %[
   .fl { color: #000000; }
   .ref { color: #000000; font-style: oblique}
   ]

   def self.add_style(css_defns)
      if css_defns[-1] != "\n" then css_defns += "\n" end
      @styles += css_defns
   end


   def self.flush(before: "", after: "", asis: false)
      if @buffer.length + @styles.length + before.length + after.length == 0
         return nil  # if there is nothing to do, do NOT open the file at all
      end
      File.open(@path, mode: 'a') do |stream|
         if @styles.length > 0
            stream.puts "<style>"
            stream.puts @styles
            stream.puts "</style>"
            @styles = ''
         end
         if @buffer.length > 0
            stream.puts "<p>Log flushed, #{Time.now}: #{@buffer.length} entries</p>\n"
         end
         if before.length > 0            
            stream.puts( if asis then before else "<p class=\"fl\">#{before}</p>" end )
         end
         if @group_by != :time then @buffer.sort! end
         stream.puts "<table>"
         @buffer.each { |msg| stream.puts msg.to_html(@count) }
         stream.puts "</table>"
         if after.length > 0
            stream.puts( if asis then after else "<p class=\"fl\">#{after}</p>" end )
         end
         stream.puts "<hr>"
      end
      @buffer = []
      nil
   end


   def self.close()
      if @path != nil
         after = "<p>Log ended, #{Time.now}: #{@count} entries in all.</p>\n</code>\n</body>"
         flush(after: after)
         @path = nil
      end
   end

   def self.finalize(exception, esc: self.esc)
      begin
         err "finalize", exception, esc
         if not @path.nil? 
            close
         end
      rescue Exception => exc 
         @path = nil
         STDERR.puts "Finalizing the log failed.\n\n#{exc}"
      end
      nil
   end        
end


class DbgMessage
   include Comparable
   attr_reader :key, :match, :index, :caller
   attr_accessor :body, :priority, :data_class
   #HTML_ESCAPER = HTMLEntities.new 
   PRIORITY_LEVELS = {
      :ERROR => 4, :WARNING => 3, :PRIORITY => 2, :ORDINARY => 1, :UNKNOWN => 0
   }
   STYLE_CLASSES = {
      :ERROR => 'er', :WARNING => 'wn', :PRIORITY => 'hp',
      :ORDINARY => 'lp', :UNKNOWN => 'un'
   }
   STYLES = %[
      .er {
         color: #ff0000; font-weight: bold; background-color: #F0D0D0;
         border: 1px solid red; margin-left: 1em;
      }
      .hp {
         color: #007020; font-weight: bold; background-color: #E1FDC8;
         border: 1px solid #007020; margin-left: 1em;
      } 
      .lp {color: #000000; background-color: #DDDDDD;
         border: 1px solid black; margin-left: 1em;
      } 
      .wn {
         color: #0000ff; font-weight: bold; background-color: #C8E5F8;
         border: 1px solid blue; margin-left: 1em;
      } 
      .un {
         color: #AEAF85; font-weight: bold; background-color: #94A2B7;
         border: 1px solid ccd847; margin-left: 1em;
      } 
   ]
   
   def self.err(index: nil, caller: nil, body: nil, style: nil)
      msg = self.new index: index, caller: caller, body: body
      msg.priority = PRIORITY_LEVELS[:ERROR]
      if style.nil? then msg.data_class = STYLE_CLASSES[:ERROR]
      else  msg.data_class = style
      end
      msg
   end
   def self.warn(index: nil, caller: nil, body: nil, style: nil)
      msg = self.new index: index, caller: caller, body: body
      msg.priority = PRIORITY_LEVELS[:WARNING]
      if style.nil? then msg.data_class = STYLE_CLASSES[:WARNING]
      else  msg.data_class = style
      end
      msg
   end
   private 
   def self.key_check(key)
      if not key.kind_of? String
         raise TypeError.new "DbgMgr key is a #{key.class}, not a String as required."
      elsif key.length == 0
         raise ArgumentError "Empty string is not an allowable DbgMgr key."
      end
   end
   public
   def self.priority(key: nil, match: nil, index: nil, caller: nil, body: nil, style: nil)
      key_check key
      msg = self.new key: key, match: match, index: index, caller: caller, body: body
      msg.priority = PRIORITY_LEVELS[:PRIORITY]
      if style.nil? then msg.data_class = STYLE_CLASSES[:PRIORITY]
      else  msg.data_class = style
      end
      msg
   end
   def self.ordinary(key: nil, match: nil, index: nil, caller: nil, body: nil, style: nil)
      key_check key
      msg = self.new key: key, match: match, index: index, caller: caller, body: body
      msg.priority = PRIORITY_LEVELS[:ORDINARY]
      if style.nil? then msg.data_class = STYLE_CLASSES[:ORDINARY]
      else  msg.data_class = style
      end
      msg
   end

   def initialize(key: nil, match: nil, index: nil, caller: '', body: '')
      @key   = key or ''
      @match = if match then match else key end
      @index = index
      @caller =  DbgMgr.html_encode caller
      @body = body
      @priority = PRIORITY_LEVELS[:UNKNOWN]
      @data_class = STYLE_CLASSES[:UNKNOWN]
   end
   def ==(other)
      @index == other.index
   end
   def !=(other)
      @index != other.index
   end
   def <=>(other)
      grouping = DbgMgr.group_by
      if grouping == :time then return @index <=> other.index # most common call is this
      elsif grouping == :caller # within caller by time, not priority
         if @caller < other.caller then return -1
         elsif other.caller < @caller then return 1
         end
      else # group by priority and, if requested, group within that by key or match
         if @priority > other.priority then return -1 
         elsif other.priority > @priority then return 1
         elsif @key.nil? and other.key.nil?
            return @index <=> other.index # same priority, only time of entry counts
         elsif @key.nil? ^ other.key.nil? # ie. one, but not both are nil... never happens
            raise "Invalid priority for #{self} relative to #{other}"
         end
         # we get here if both have the same priority and both have keys,
         # and thus also have non-nil "match" attributes
         if grouping == :key 
            if @key < other.key then return -1
            elsif other.key < @key then return 1
            end
         elsif grouping == :match
            if @match < other.match then return -1
            elsif other.match < @match then return 1
            end
         end
      end
      @index <=> other.index
   end      
   def to_html(largest_index)
      index_width = 1 + Integer(Math.log10 largest_index)
      formatted_index = index.to_s 
      while formatted_index.length < index_width
         formatted_index = "&nbsp;"+formatted_index
      end
      key = if @key.nil? then '' else "[#{@key}]" end
      source_cell  = "<td class=\"ref\">#{formatted_index}: #{@caller}#{key}</td>\n"
      data_cell = "<td class=\"#{@data_class}\">#{body}</td>"
      "<tr>#{source_cell}#{data_cell}</tr>"
   end
end

File: libpy/lib/dbg.py

from blist import blist
import html
import inspect
import io
import os
from singleton import OneOnly
from sys import stdout, stderr
import sysutils as su
import time
import traceback
from wildcardlookup import WildCardLookup

def getDbgMgr(*patterns):
   mgr = DbgMgr.get()
   mgr.add_patterns(*patterns)
   return mgr
        

def openDbgMgr(*patterns, **kwargs): 
   mgr = DbgMgr.get()
   mgr.add_patterns(patterns)
   if mgr.path is None:
      path = su.getitem(kwargs, "path", None)
      mgr._open(path)
      mgr.sortby = su.getitem(kwargs, "sortby", "time")
   return mgr
   

class DbgMessage:
   DEFAULTS = {
      "key": "",              # not used by errors and warnings
      "weakest_match": None,  # weakest pattern that matches the key (if key is non-empty)
      "display_priority": 3,  # the priority assigned to ordinary debugging output
      "caller": '??',         # should always be supplied by the caller
      "body": "",             # should always be supplied by the caller
      "index": -1,            # should always be supplied by the caller
      "style_class": "ordinary_msg" # ordinary, non-priority debugging output
   }
   def __init__(self, parms_in):
      parms = su.mergepairs(parms_in, self.DEFAULTS, clone=True)
      self.key = parms['key']
      self.match = parms['weakest_match'] if parms['weakest_match'] else parms['key']
      self.display_priority = parms['display_priority']
      self.caller = parms['caller']
      self.body = parms['body']
      self.index = parms['index']
      self.style_class = parms['style_class']

   def __eq__(self, other):
      self.index is other.index

   def __ne__(self, other):
      not (self.index is other.index)

   def __lt__(self, other):
      # the display priority always rules:
      if self.display_priority < other.display_priority:
         return True
      elif self.display_priority is other.display_priority:
         if self.key is None: 
            # both are either an error or a warning, so the index rules
            return self.index < other.index
         else:
             # Both must be debugging messages, since they have the same priority and
             # are not errors or warnings. They should both, therefore, have keys,
             # weakest matches, and sources.
            pivot = DbgMgr.get().sortby
            if pivot == "key":
               if self.key < other.key:
                  return True
               elif self.key == other.key:
                  if self.index < other.index:
                     return True
            elif pivot == "match":
               if self.match < other.match: 
                  return True
               elif self.match == other.match:
                  if self.index < other.index:
                     return True
            elif pivot == "source":
               if self.source < other.source:
                  return True
               elif self.source == other.source:
                  if self.index < other.index:
                     return True
      return False

   def __le__(self, other):
      return self.__lt__(other) or self.__eq__(other)

   def __gt__(self, other):
      return not (self.__lt__(other) or self.__eq__(other))

   def __ge__(self, other):
      return not self.__lt__(other)

   def __str__(self):
      from_key = "[{}]".format(self.key) if self.key else ""
      clean_caller =  html.escape(self.caller)
      about = "<span class=\"msg_source\">{0}: {1}{2}</span>".format(
         self.index, clean_caller, from_key
      )
      table = "<table class=\"{0}\"><tr><td>{1}</td></tr></table>".format(
         self.style_class, self.body
      )
      return about+table


@OneOnly
class DbgMgr:
   def __init__(self):
      self.buffer = []
      self.count = 0
      self.highpriority = WildCardLookup()
      self.keypatterns = WildCardLookup()
      self.message_styles = {
         "error_msg": """{color: #ff0000; font-weight: bold; background-color: #F0D0D0; 
                   border: 1px solid red; margin-left: 1em;}""",
         "priority_msg": """{color: #007020; font-weight: bold; background-color: #E1FDC8;
                   border: 1px solid #007020; margin-left: 1em;}""", 
         "ordinary_msg": """{color: #000000; background-color: #DDDDDD; 
                   border: 1px solid black; margin-left: 1em;}""", 
         "warning_msg": """{color: #0000ff; font-weight: bold; background-color: #C8E5F8;
                   border: 1px solid blue; margin-left: 1em;}""", 
         "flush_before_after":  "{color: #000000;}", 
         "msg_source": "{color: #000000; font-style: oblique}"
      }
      self._path = None
      self._sortby = None
      self._stream = None
      self._head = None
      self._head_started = False
      self._head_is_open = False
      self._something_written = False
      self._bail_on_warning = False

   def get_path(self):
      return self._path

   def set_path(self, logdirpath):
      if self._path is None:
         abspath = self._normalizeThePath(logdirpath)
         if not os.path.exists(abspath):
            os.makedirs(abspath)
         elif not os.path.isdir(abspath):
            msg = "'{0}' exists, but is not a directory".format(abspath)
            raise NotADirectoryError(msg)
         self._path = os.path.join(abspath, "dbg"+su.now4FileName()+".html")
      elif logdirpath is None:
         self.stream = None # flushes and closes the stream
         self._path = None
      else:
         raise RuntimeError("DbgMgr already opened '{}'".format(self._path))

   def _normalizeThePath(self, path):
      rootpath = os.getcwd()
      if path:
         if os.path.isabs(path):         
            raw = path
         elif path.startswith("./") or path.startswith("../") or path.startswith("~/"):
            raw = path # implied absolute paths: "here" and "home"                 
         else:
            raw =  os.path.join(rootpath, path)
      else:
         raw = os.path.join(rootpath, "log")
      return os.path.abspath(raw) # gets rid of pesky dots, double dots and tildes

   def del_path(self):
      raise RuntimeError("Cannot remove the path attribute")

   def get_sortby(self):
      return self._sortby

   def set_sortby(self, new_value):
      new_value = new_value.lower()
      if new_value in ["key", "match", "source", "time"]:
         self._sortby = new_value
      elif new_value is None:
         self._sortby = "time"
      else:
         raise ValueError("Unrecognized sort type: '{}'".format(new_value))

   def del_sortby(self):
      raise RuntimeError("Cannot remove the sortby attribute")

   def get_stream(self):
      return stderr if self._stream is None else self._stream

   def set_stream(self, astream):
      if self._stream is None:
         self.dbg("dbg", "initializing stream: {}".format(astream))
         self._stream = astream
      else:
         self.dbg("dbg", "resetting stream to None")
         if astream in [stdout, stderr, None]:
            if self._stream in [stdout, stderr]:
               if astream != self._stream:
                  self.flush()
                  self._stream = astream
               # else: no change: output is already going to "astream"
            else:
               # closing the current output file flushes the buffer
               self.close()
               self._stream = astream
         else:
            raise RuntimeError("Output stream is already opened")

   def del_stream(self):
      raise RuntimeError("Cannot remove the stream attribute")

   path   = property(get_path, set_path, del_path)
   sortby = property(get_sortby, set_sortby, del_sortby)
   stream = property(get_stream, set_stream, del_stream)

   def mycallersname(self): 
      return inspect.currentframe().f_back.f_back.f_code.co_name

   def prepare_patterns(self, patterns):
      as_blist = blist(patterns)
      su.flatten(as_blist)
      extras = []
      for n in range(0, len(as_blist)):
         entry_n = as_blist[n]
         if ',' in entry_n:            
            actual_entries = [s.strip() for s in entry_n.split(',')]
            as_blist[n] = actual_entries[0] # replace the comma list with its first entry
            extras += actual_entries[1:]    # do NOT change the size of as_blist
         else:
            as_blist[n] = entry_n.strip()
      as_blist += extras
      as_blist.sort()
      return as_blist

   def addstyleclass(self, name, spec):
      # I try to keep all the style classes together in the head of the document, so I buffer them
      # until the first flush.  Once they are written, I write a new <style> element for each new
      # class.  I do insist that a class be declared only once or with the same spec each time. 
      #
      # The reason for having this call is to allow one to create new message types easily, not to
      # encourage your artistic genius.  Simple classes only, please!
      if name.startswith("."): name = name[1:]
      if name in self.message_styles:
         if spec != self.message_styles[name]:
            raise ValueError("Debugging style class '"+name+"' mulitply declared.")
      else:
         self.message_styles[name] = spec
         if self._head_started:
            as_html = "<style type=\"text/css\">.{0} {1}</style>\n".format(name, spec)
            if self._head_is_open:
               self._head += as_html
            else:
               self.stream.write(as_html)

   def __enter__(self):
      return self

   def __exit__(self, exc_type, exc_value, exc_tb):
      if exc_type:
         tb = io.StringIO()
         traceback.print_tb(exc_tb, limit=10, file=tb)
         self.err("{0}: {1}\n{2}".format(exc_type, exc_value, tb.getvalue()))
      self.close()
   
   def __contains__(self, name):
      return name in self.keypatterns

   def __getitem__(self, what):
      if type(what) is int:
         return self.buffer[what]
      elif type(what) is str:
         whatMatcher = WildCardLookup(what)
         return [self.buffer[x] for x in self.buffer if self.buffer[x].key in whatMatcher]
      if type(what) is slice:
         what = su.slice2range(what, max=len(self.buffer))
      if type(what) is range:
         return [self.buffer[x] for x in what]
      else:
         msg = "{0} cannot be used to index {1}"
         raise TypeError(msg.format(su.a_classname(what), su.a_classname(self)))

   def __iter__(self): return self.keypatterns.__iter__()

   def pattern_iterator(self): return self.keypatterns.__iter__()
   def buffer_iterator(self): return self.buffer.__iter__()

   def pattern_count(self): return len(self.keypatterns)
   def msg_count(self): return self.count - 1 # message numbering starts at 1
   def buffer_count(self): return len(self.buffer)

   def _open(self, logdirpath):
      if self.path:
         return self
      if self.stream in [stdout, stderr]: 
         self.stream = None
      elif self.stream:
         self.stream.close()
         self.stream = None
      self.path = logdirpath 
      return self._start_the_head()

   def _start_the_head(self):
      headlines = [
         '<!DOCTYPE html>'
         '<html>'
         '<head>'
         '<meta charset="utf-8">',
         '<style type="text/css">'
         ]
      self._head_started = True
      self._head_is_open = True
      lines = [
      "."+name+" "+self.message_styles[name] for name in self.message_styles
      ]
      lines.sort() # NB. sort does NOT return "self"
      headlines += lines
      headlines.append('</style>\n')
      self._head = "\n".join(headlines)
      return self

   def add_patterns(self, *new_patterns):
      for pattern in self.prepare_patterns(new_patterns):
         if len(pattern) > 0:
            if pattern[0] == '+':
               pattern = pattern[1:]
               self.highpriority.add(pattern)
            self.keypatterns.add(pattern)

   def remove_patterns(self, *dead_patterns):
      for pattern in self.prepare_patterns(dead_patterns):
         self.keypatterns.remove(pattern)
         self.highpriority.remove(pattern)

   def raise_priority(self, *ptns):
      for ptn in self.prepare_patterns(ptns):
         self.keypatterns.add(ptn)
         self.highpriority.add(ptn)

   def lower_priority(self, *ptns):
      for ptn in self.prepare_patterns(ptns):
         self.highpriority.remove(ptn)

   def all_matched(self, *keys):
      return self.keypatterns.all_matched(self.prepare_patterns(keys))

   def any_matched(self, *keys):
      return self.keypatterns.any_matched(self.prepare_patterns(keys))

   def buffer_message(self, parms):
      self.count += 1
      parms["index"] = self.count
      self.buffer.append(DbgMessage(parms))

   def dbg(self, key, text, *, src=None, style=None, now=False):
      if (key in self.keypatterns) or not key:
         parms =  {}
         parms["key"] = key or ""
         parms["weakest_match"] = self.keypatterns.weakestPatternFor(key)
         parms["caller"] = self.mycallersname() if src is None else src
         parms["display_priority"] = 3
         if style: parms["style_class"] = style
         elif key in self.highpriority:  parms["style_class"] = "priority_msg"
         else: parms["style_class"] = "ordinary_msg"
         parms["body"] = text
         self.buffer_message(parms)
         if now: 
            if isinstance(now, str): self.flush(after=now)
            else: self.flush()
      # else key is not active, so I do nothing

   def err(self, text_or_err, src=None, bail_out=True):
      parms =  {}
      parms["caller"] =  self.mycallersname() if src is None else src
      parms["body"] = "<pre>"+str(text_or_err)+"</pre>"
      parms["display_priority"] = 1   # in English: "first priority!"
      parms["style_class"] = "error_msg"
      self.buffer_message(parms)
      if bail_out:
         if isinstance(text_or_err, Exception):
            raise text_or_err
         else:
            raise RuntimeError(text_or_err)

   def pre(self, key, text, *, src=None, style=None, now=False):
      if (key in self.keypatterns) or not key:
         if not src:
            src = self.mycallersname()
         self.dbg(key, "<pre>" + text + "</pre>", src=src, style=style, now=now)

   def warn(self, text_or_err, src=None, bail_out = False):
      parms =  {}
      parms["caller"] =  self.mycallersname() if src is None else src
      parms["body"] = "<pre>"+str(text_or_err)+"</pre>"
      parms["display_priority"] = 2   # second priority: after errors
      parms["style_class"] = "warning_msg"
      self.buffer_message(parms)
      if (bail_out or self._bail_on_warning):
         if isinstance(text_or_err, Exception):
            raise text_or_err
         else:
            raise RuntimeError(text_or_err)

   def flush(self, before="", after="", asis=False):
      self.dbg("dbg", "flush() called by {}".format(self.mycallersname()))
      if self.buffer_count() is 0 and len(before) is 0 and len(after) is 0:
         return 
      if not self._head_started:
         self._start_the_head()
      if self._head_is_open: # first time switch: close the head and open the body
         self.stream = open(self.path, mode='x')
         self.stream.write(self._head)
         self.stream.write('</head>\n<body>\n<code>\n')
         self._head_is_open = False
      if self.sortby is not "time":
         self.buffer.sort()
      if len(before) > 0:
         if asis:
            self.stream.write(before)
         else:
            self.stream.write('<p class="flush_before_after">{0}</p>\n'.format(before))
      self.dbg("dbg", "flushing buffer to {}".format(self.stream))
      for entry in self.buffer:
         self.stream.write(str(entry))
      if len(after) > 0:
         if asis:
            self.stream.write(after)
         else:
            self.stream.write('<p class="flush_before_after">{0}</p>\n'.format(after))
      self.stream.write("<hr>\n")
      self.buffer = []
      self._something_written = True

   def close(self, before="", after="", asis=False):
      self.flush(before, after, asis)
      if self._something_written:
         msg = "<p>Log ended, {1}: {0} entries</p>\n".format(self.count, time.asctime())
         self.stream.write(msg)
         self.stream.write('</code>\n</body>\n')
         if self._stream not in [stdout, stderr, None]:
            self._stream.close()
         elif self._stream is not None:
            self._stream.flush() # dump the stream's buffer (ours is empty!!!)
         self._stream = None
